wikithingsdb/fetch.py

```diff
@@ -0,0 +1,72 @@
+from wikithingsdb.engine import engine
+from wikithingsdb.models import Page, Redirect, WikiClass, Type, DbpediaClass
+from sqlalchemy.orm import sessionmaker
+from sqlalchemy import func
+
+session = sessionmaker(bind=engine)()
+
+
+# def _clean_query(query):
+#     query = query.replace(' ', '_')
+#     query = query.lower()
+#     return query
+
+
+# What are all the ways to refer to ARTICLE?
+def types_of_article(article):
+    # article = _clean_query(article)
+    result = session.query(Page).\
+        join(Page.types).\
+        filter(Page.page_title == article).one()
+    return [x.type for x in result.types]
+
+
+def classes_of_article(article):
+    result = session.query(Page).\
+        join(Page.classes).\
+        filter(Page.page_title == article).one()
+    return [x.class_name for x in result.classes]
+
+
+def hypernyms_of_article(article):
+    return {w_class: hypernyms_of_class(w_class)
+            for w_class in classes_of_article(article)}
+
+
+# also
+def hypernyms_of_class(w_class):
+    result = session.query(WikiClass).\
+        join(WikiClass.dbpedia_classes).\
+        filter(WikiClass.class_name == w_class).one()
+    return [x.dpedia_class for x in result.dbpedia_classes]
+
+
+# What are all the articles of TYPE, CLASS, DBPEDIA_CLASS?
```

Comment:
Actually, they are quite fast. `fetch.articles_of_hypernym('owl:Thing')` took about a second or so (for 678 articles). Boris specifically asked for this "backwards" relationship and Sue liked the idea. The only function that is noticeably slow (but sporadic) is `fetch.redirects_of_article()`, I think because those tables are not pre-indexed by sqlalchemy.